src/main/webapp/app/markdown-editor/markdown-editor.component.ts

```diff
@@ -0,0 +1,157 @@
+import {Command, BoldCommand, ItalicCommand, UnderlineCommand} from 'app/markdown-editor/commands';
+import {ArtemisMarkdown} from 'app/components/util/markdown.service';
+import {SpecialCommand} from 'app/markdown-editor/specialCommands';
+
+@Component({
+    selector: 'jhi-markdown-editor',
+    styleUrls: ['./markdown-editor.scss'],
+    providers: [ArtemisMarkdown],
+    templateUrl: './markdown-editor.component.html'
+})
+export class MarkdownEditorComponent implements AfterViewInit, OnInit {
+    @ViewChild('aceEditor')
+    aceEditorContainer: AceEditorComponent;
+    aseEditorOptions = {
+        autoUpdateContent: true,
+        mode: 'markdown',
+    };
+
+    @Input() defaultText: string;
+    @Input() specialCommands: SpecialCommand[];
+    @Input() useMarkdownPreview: boolean;
+
+    @Output() html = new EventEmitter<string>();
+    @Output() textWithSpecialCommandFound = new EventEmitter<[string, SpecialCommand]>();
+    @Output() previewModeChange = new EventEmitter<boolean>();
+
+    @ContentChild('preview') previewChild: ElementRef;
+
+
+    previewTextAsHtml: string;
+    previewMode = false;
+    defaultCommands: Command[] = [new BoldCommand(), new ItalicCommand(), new UnderlineCommand()];
+
+    constructor(private artemisMarkdown: ArtemisMarkdown) {}
+
+    get previewButtonTranslateString(): string {
+        return this.previewMode ? 'entity.action.edit' : 'entity.action.preview';
+    }
+
+    get previewButtonIcon(): string {
+        return this.previewMode ? 'pencil-alt' : 'eye';
+    }
+
+    get showDefaultPreview(): boolean {
+        return this.previewChild == null;
+    }
+
+    addCommand(command: Command) {
+        this.defaultCommands.push(command);
+    }
+
+    removeCommand(classRef: typeof Command) {
+        setTimeout(() =>
+            this.defaultCommands = this.defaultCommands.filter(element => !(element instanceof classRef))
+        );
+    }
+
+    ngAfterViewInit(): void {
+        this.setupMarkdownEditor();
+    }
+
+    ngOnInit(): void {
+        [...this.defaultCommands, ...this.specialCommands || []].forEach(command => {
+            command.setEditor(this.aceEditorContainer.getEditor());
+        });
+    }
+
+    /**
+     * @function setupQuestionEditor
+     * @desc Initializes the ace editor for the mc question
+     */
+
+    /** Currently responsible for making the editor appear nicely**/
+    setupMarkdownEditor(): void {
+        this.aceEditorContainer.setTheme('chrome');
+        this.aceEditorContainer.getEditor().renderer.setShowGutter(false);
+        this.aceEditorContainer.getEditor().renderer.setPadding(10);
+        this.aceEditorContainer.getEditor().renderer.setScrollMargin(8, 8);
+        this.aceEditorContainer.getEditor().setHighlightActiveLine(false);
+        this.aceEditorContainer.getEditor().setShowPrintMargin(false);
+        this.aceEditorContainer.getEditor().clearSelection();
+        this.aceEditorContainer.getEditor().on(
+            "change",
```

Comment:
If we really need to parse on change, we can discard all the other logic on when to parse the content.
What this currently does is calling parse on every change. The client cannot handle this right now, as it does not know when a new parsing cycle starts.
Also, this leads to lots of unnecessary parsing.

So what I am saying is that we:
1. should think twice if we want to do these automatic parsing cycles, or
1. allow the client to handle this case (e.g. by emitting our own onChange Event), or
1. require to enable this behaviour.

In most (if not all) cases, this will **not** be necessary.